### PR TITLE
[MNT-24127] Added Endpoint to Calculate the Folder Size.

### PR DESCRIPTION
**Pull request by mohit-singh4**
https://hyland.atlassian.net/browse/MNT-24127

### CODE REVIEW

**Review by suneet-gupta — APPROVED** [review]
I have already done the review as part of https://github.com/Alfresco/rest-api-explorer/pull/213 `but somehow PR was accidentally merged without Architect approval. Approving the PR based on previous comments and changes.

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 1]
```diff
@@ -2484,7 +2491,96 @@ paths:
...
+      summary: Get the size of folder after initiating POST/calculateSize Endpoint.
+      description: |
+        ***Receiving the size***
+        
+        This endpoint takes ```nodeId``` as a path parameter and returns the result after executing above POST request.
+        
+        Below are the three stages, which reflect the status for the present execution ***nodeId***.
+        
+        ```
+           NOT INITIATED: Not Iniated yet.
```
Comment (kavitshah-gl) [inline src/main/webapp/definitions/alfresco-core.yaml:2557]: Don't have space in the status.

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 2]
```diff
@@ -10690,6 +10786,30 @@ definitions:
...
+      numberOfFiles:
+        type: integer
+        description: Provides the number of files in the folder.
+      status:
+        type: string
+        description: Provides a status that signifies the completion of the calculateSize mechanism.
+  sizeAcknowledgedResponse:
+    type: object
+    properties:
+      executionId:
```
Comment (kavitshah-gl) [inline src/main/webapp/definitions/alfresco-core.yaml:10810]: This should be NodeID rather than execution ID.

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 1 above)
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml]: Done

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 3]
```diff
@@ -123,12 +123,19 @@ parameters:
     description: The identifier of a target node.
     required: true
     type: string
-  nodeFolderIdParam:
```
Comment (tpage-alfresco) [inline src/main/webapp/definitions/alfresco-core.yaml:126]: Did you intend to delete this? It seems unrelated to your changes.

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 4]
```diff
@@ -123,12 +123,19 @@ parameters:
...
-    required: true
-    type: string
+  maxItemsForSizeParam:
+    name: maxItems
+    in: query
+    description: |
+      This assists in handling scenarios where ***large files*** or ***subfolders*** exist within a folder,
+      allowing the folder size determination mechanism to function in chunks.
+      
+      So that calculating the huge folder will be done efficiently.
```
Comment (tpage-alfresco) [inline src/main/webapp/definitions/alfresco-core.yaml:133]: This is a fragment of a sentence - could you expand it into a full sentence so that it's clear what you're referring to?

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 5]
```diff
@@ -123,12 +123,19 @@ parameters:
...
-    type: string
+  maxItemsForSizeParam:
+    name: maxItems
+    in: query
+    description: |
+      This assists in handling scenarios where ***large files*** or ***subfolders*** exist within a folder,
+      allowing the folder size determination mechanism to function in chunks.
+      
+      So that calculating the huge folder will be done efficiently.  
+      If not supplied then the default value is 100.
```
Comment (tpage-alfresco) [inline src/main/webapp/definitions/alfresco-core.yaml:134]: Is there a server-side maximum too? Otherwise I think this provides a DoS method for authenticated users.

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 6]
```diff
@@ -2484,7 +2491,96 @@ paths:
...
+        This may be executed against a node specified by **nodeId** and this nodeId must be a valid node ID.
+        Furthermore, this POST endpoint will be executed **asynchronously** with a `202` HTTP response signifying that
+        the request has been accepted successfully.
+        
+        The response body will include the ```nodeId``` of the pending action, which can be used in a
+        **GET/calculateSize** endpoint to check if the action's status has been completed, at which point the result will be returned, comprising ```the size of the node in bytes```.
+        
+        Essentially, ```maxItems``` ***parameter*** is utilized for pagination purposes.
+        
+        By default, **results are limited to the first 100.**
```
Comment (tpage-alfresco) [inline src/main/webapp/definitions/alfresco-core.yaml:2510]: I read the documentation above and here, and I still don't know what will happen if I set `maxItems`. Will I get the size of the first `maxItems` in the folder, or will it be an estimate of the total size based on the first `maxItems`?

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 7]
```diff
@@ -2484,7 +2491,96 @@ paths:
...
+        Results can be restricted using "paging". For example:
```
Comment (tpage-alfresco) [inline src/main/webapp/definitions/alfresco-core.yaml:2511]: You don't mention `paging` or `skipCount` in the parameters list below.

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 8]
```diff
@@ -2484,7 +2491,96 @@ paths:
...
+          description: Request Accepted
+          schema:
+            $ref: '#/definitions/sizeAcknowledgedResponse'
+        '400':
+          description: |
+            Invalid parameter: value of **nodeId** is invalid
+        '401':
+          description: Authentication failed
+        '403':
+          description: Current user does not have permission for **nodeId**
```
Comment (tpage-alfresco) [inline src/main/webapp/definitions/alfresco-core.yaml:2538]: What happens if the current user has access to `nodeId` but not to a child folder?

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 9]
```diff
@@ -2484,7 +2491,96 @@ paths:
...
+    get:
+      tags:
+        - nodes
+      summary: Get the size of folder after initiating POST/calculateSize Endpoint.
+      description: |
+        ***Receiving the size***
+        
+        This endpoint takes ```nodeId``` as a path parameter and returns the result after executing above POST request.
+        
+        Below are the three stages, which reflect the status for the present execution ***nodeId***.
```
Comment (tpage-alfresco) [inline src/main/webapp/definitions/alfresco-core.yaml:2554]:
```suggestion
        The size calculation process has three stages:
```

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 10]
```diff
@@ -2484,7 +2491,96 @@ paths:
...
+        
+        ```
+           NOT-INITIATED: Not Initiated yet.
```
Comment (tpage-alfresco) [inline src/main/webapp/definitions/alfresco-core.yaml:2557]: Is it possible to receive this?

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 11]
```diff
@@ -2484,7 +2491,96 @@ paths:
...
+           IN-PROGRESS:   Calculating execution is in progress.
```
Comment (tpage-alfresco) [inline src/main/webapp/definitions/alfresco-core.yaml:2558]:
```suggestion
           IN-PROGRESS:   Calculation is in progress.
```

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 12]
```diff
@@ -2484,7 +2491,96 @@ paths:
...
+           COMPLETED:     Calculation has been done.
```
Comment (tpage-alfresco) [inline src/main/webapp/definitions/alfresco-core.yaml:2559]: Did you look to see if there's an existing standard for upper or lower case keywords in our REST API?

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 13]
```diff
@@ -2484,7 +2491,96 @@ paths:
...
+        '404':
+          description: Entity with **nodeId** does not exist
+        default:
+          description: Unexpected error
+          schema:
+            $ref: '#/definitions/Error'
+    get:
+      tags:
+        - nodes
+      summary: Get the size of folder after initiating POST/calculateSize Endpoint.
```
Comment (tpage-alfresco) [inline src/main/webapp/definitions/alfresco-core.yaml:2548]:
```suggestion
      summary: Get the result of a background job to calculate the size of a folder.
```

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 14]
```diff
@@ -10690,6 +10786,30 @@ definitions:
         $ref: '#/definitions/PermissionsInfo'
       definition:
         $ref: '#/definitions/Definition'
+  calculateSizeResponse:
+    type: object
+    properties:
+      id:
+        type: string
+        description: Provides the id for which POST/calculateSize has been initiated.
```
Comment (tpage-alfresco) [inline src/main/webapp/definitions/alfresco-core.yaml:10794]: This is a little bit weird. For consistency I think this should be the id of the background job, not the node. If the background job does not have an id then this is ok.

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 15]
```diff
@@ -10690,6 +10786,30 @@ definitions:
...
+      size:
+        type: string
+        description: Provides a folder size in bytes.
+      calculatedAtTime:
+        type: string
+        description: Provides the time when the calculating folder size will be completed.
```
Comment (tpage-alfresco) [inline src/main/webapp/definitions/alfresco-core.yaml:10800]: Do you mean this?

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 16]
```diff
@@ -10690,6 +10786,30 @@ definitions:
...
+  calculateSizeResponse:
+    type: object
+    properties:
+      id:
+        type: string
+        description: Provides the id for which POST/calculateSize has been initiated.
+      size:
+        type: string
+        description: Provides a folder size in bytes.
+      calculatedAtTime:
```
Comment (tpage-alfresco) [inline src/main/webapp/definitions/alfresco-core.yaml:10798]:
```suggestion
      calculatedAt:
```

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 17]
```diff
@@ -10690,6 +10786,30 @@ definitions:
...
+        type: string
+        description: Provides the time when the calculating folder size will be completed.
+      numberOfFiles:
+        type: integer
+        description: Provides the number of files in the folder.
+      status:
+        type: string
+        description: Provides a status that signifies the completion of the calculateSize mechanism.
```
Comment (tpage-alfresco) [inline src/main/webapp/definitions/alfresco-core.yaml:10806]: This would be a better place to include the list of possible states.

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 18]
```diff
@@ -10690,6 +10786,30 @@ definitions:
...
+  sizeAcknowledgedResponse:
+    type: object
+    properties:
+      nodeId:
+        type: string
+        description: Provides nodeId indicating that the request has been acknowledged.
```
Comment (tpage-alfresco) [inline src/main/webapp/definitions/alfresco-core.yaml:10812]: I'd expect the response to contain the id of the background job.

**Review by tpage-alfresco — CHANGES_REQUESTED** [review]
I had quite a few comments/suggestions.

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 19]
```diff
@@ -2484,7 +2491,96 @@ paths:
...
+  '/nodes/{nodeId}/calculateSize':
+    post:
+      tags:
+        - nodes
+      summary: Calculate a folder size
+      description: |
+        
+        This may be executed against a node specified by **nodeId** and this nodeId must be a valid node ID.
+        Furthermore, this POST endpoint will be executed **asynchronously** with a `202` HTTP response signifying that
+        the request has been accepted successfully.
```
[the remark this answers is not in the document]
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml:2499]: Again, it will return the **nodeId** for which the Action was initiated and start the execution asynchronously.
Similarly, we have a **GET endpoint** to check the status and output for the specified Node.

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 8 above)
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml:2523]: As we are executing an Alfresco FTS query through the Search Subsystem, we are only receiving all linked **nodeIds** at once. As a result, we are just restricting permission validation to the provided node.

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 10 above)
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml:2542]: Yes

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 14 above)
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml:10779]: We provide the **nodeId** in the POST Endpoint response instead **backgroundJobId** because it's removed once the action is completed. As a result, we cannot discover the size in the GET Endpoint response.

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 18 above)
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml:10803]: We provide the **nodeId** in the POST Endpoint response instead **backgroundJobId** because it's removed once the action is completed. As a result, we cannot discover the size in the GET Endpoint response.

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 12 above)
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml:2544]: Yes, in /renditions endpoint we were maintaining status as **CREATED**.

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 19 above)
[the remark this answers is not in the document]
Comment (tpage-alfresco) [reply src/main/webapp/definitions/alfresco-core.yaml:2505]: I think you should add some documentation for this case. If the POST is called twice then I would find it surprising that the same id is returned both times, and I'm not clear which result I will get when calling the GET (e.g. Is the existing job cancelled? Does the new job overwrite the output of the first job?)

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 8 above)
Comment (tpage-alfresco) [reply src/main/webapp/definitions/alfresco-core.yaml:2529]: Ok - so you should either update the documentation to say:
* The calculation is run as system and so will include the size of documents that the user does not have access to.
OR
* The calculation is restricted to documents that the user has access to and so the true folder size may be larger than returned.

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 10 above)
Comment (tpage-alfresco) [reply src/main/webapp/definitions/alfresco-core.yaml:2548]: I don't see how it is possible? Doesn't the job start immediately when the user makes the POST request?

Comment (tpage-alfresco) [reply src/main/webapp/definitions/alfresco-core.yaml:2548]: Please document the answer

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 12 above)
Comment (tpage-alfresco) [reply src/main/webapp/definitions/alfresco-core.yaml:2550]: Brilliant - thanks! :+1:

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 6 above)
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml]: To efficiently retrieve and consolidate size of large, nested data, we are using **maxItems** parameter to divide the data into segments.

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 14 above)
Comment (tpage-alfresco) [reply src/main/webapp/definitions/alfresco-core.yaml:10785]: The pattern in most places in the v1 API is:

* `POST /parents/{parent-id}/children` returns `{"id": "child-id", ...}`
* `GET /parents/{parent-id}/children` returns paged response including `{"id": "parent-id", ...}`
* `GET /parents/{parent-id}/children/{child-id}` returns `{"id": "parent-id", ...}`

Here instead we are doing:

* `POST /parents/{parent-id}/action` returns `{"id": "parent-id", ...}`
* `GET /parents/{parent-id}/action` returns `{"id": "parent-id", ...}` (not paged)

Maybe this is consistent with `/nodes/{node-id}/move` and `/nodes/{node-id}/copy` though, where we're really performing an action on the resource, rather than a creation.

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 4 above)
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml]: We've refined the preceding message in single sentence.

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 17 above)
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml]: We've refined the preceding message along with possible states.

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 16 above)
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml]: Done

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 15 above)
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml]: We meant to provide the timestamp in response when the calculation of the folder size is completed.
We've refined the message.

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 5 above)
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml]: We will set a maximum value of **1000** on the server side for partitioning.
We started working on it.

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 9 above)
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml]: We've refined the preceding message.
Done

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 10 above)
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml:2548]: Previously, we configured the nodeService property attribute to handle all three stages of POST activity.
However, now we will look at the implementation with an **execution job ID** instead of nodeId and will return the **execution job ID**  in response.

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 19 above)
[the remark this answers is not in the document]
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml:2505]: Previously, we used the nodeService property attribute to store the result output.
But now we'll look at the implementation with an **execution job ID** instead of a node ID and return the **execution Job ID** as a response.

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 14 above)
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml:10785]: We are also referring to the following V1 format:
However, in the case of POST, we return the parent-id rather than the **execution job ID.**

- `POST /nodes/{node-id}/calculateSize` returns {"id": "child-id",...}.

For GET, we return a response with the parent-id (not paged).

- `GET /nodes/{node-id}/calculateSize` returns {"id": "parent-id",...}.

Note: In the preceding comment, **"action"** referred to the Action executor used to execute tasks asynchronously, not any endpoint. For better clarity and understanding, we may set up a call.

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 13 above)
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml]: We've refined the preceding summary.

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 20]
```diff
@@ -10690,6 +10786,30 @@ definitions:
...
         $ref: '#/definitions/Definition'
+  calculateSizeResponse:
+    type: object
+    properties:
+      id:
+        type: string
+        description: Provides the id for which POST/calculateSize has been initiated.
+      size:
+        type: string
+        description: Provides a folder size in bytes.
```
[the remark this answers is not in the document]
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml]: We've refined the description message.

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 10 above)
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml:2549]: Now, we are returning **executionJob** Id and maintaining all stages that correspond to that node ID.

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 19 above)
[the remark this answers is not in the document]
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml:2506]: If the POST is called twice then I would find it surprising that the same id is returned both times, and I'm not clear which result I will get when calling the GET (e.g. Is the existing job cancelled? Does the new job overwrite the output of the first job?)
 [Updated Remarks]: Now, we are considering executionId instead of nodeId. Thus, existing job will not be cancelled if we triggered POST twice [ as its working with different execution Id ].
 And the same will not overwrite the output of the first job.

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 18 above)
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml:10810]: Now, we are considering a different execution ID instead of a node ID.

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 14 above)
Comment (mohit-singh4) [reply src/main/webapp/definitions/alfresco-core.yaml:10786]: Now, we are considering an execution job ID instead of a node ID and returning the execution job ID as a response in POST endpoint.

**Review by pzhyland — APPROVED** [review]
Approved but please focus on providing the **E2E test**.
